cmd/object-handlers-common.go

```diff
@@ -264,3 +271,100 @@ func setPutObjHeaders(w http.ResponseWriter, objInfo ObjectInfo, delete bool) {
 		}
 	}
 }
+
+// ilmLimitNoncurrentVersions limits the number of noncurrent versions of an
+// object based on lifecycle rules configured on the bucket.
+func ilmLimitNoncurrentVersions(ctx context.Context, objectAPI ObjectLayer, bucket, object string) {
+	if enable := env.Get(api.EnvAPIILMInline, ""); enable == "" {
+		return
+	}
+
+	var lc *lifecycle.Lifecycle
+	var err error
+	if lc, err = globalLifecycleSys.Get(bucket); err != nil {
+		if errors.Is(err, BucketLifecycleNotFound{Bucket: bucket}) {
+			return
+		}
+		logger.LogIf(ctx, fmt.Errorf("failed to fetch bucket lifecycle: %w", err))
+		return
+	}
+
+	objOpts := lifecycle.ObjectOpts{Name: object}
+	lim := lc.NoncurrentVersionsExpirationLimit(objOpts)
+	if lim == 0 { // no limit set
+		return
+	}
+
+	var (
+		noncurrentVersions []ObjectToDelete
+		keep               int
+	)
+	versioned := globalBucketVersioningSys.Enabled(bucket)
+	results := make(chan ObjectInfo)
+	err = objectAPI.Walk(ctx, bucket, object, results, ObjectOptions{WalkVersions: versioned})
+	if err != nil {
+		logger.LogIf(ctx, fmt.Errorf("failed to walk %s/%s: %v", bucket, object, err))
+		return
+
+	}
+
+	rcfg, _ := globalBucketObjectLockSys.Get(bucket)
+	for oi := range results {
+		if oi.IsLatest {
+			continue
+		}
+
+		keep++
+		if keep <= lim {
+			continue
+		}
+		if rcfg.LockEnabled {
+			locked := enforceRetentionForDeletion(ctx, oi)
+			if locked {
+				continue
+			}
+		}
+		noncurrentVersions = append(noncurrentVersions, ObjectToDelete{
+			ObjectName: oi.Name,
+			VersionID:  oi.VersionID,
+		})
+	}
+
+	deleteObjectVersions(ctx, objectAPI, bucket, noncurrentVersions)
+}
+
+func deleteObjectVersions(ctx context.Context, o ObjectLayer, bucket string, toDel []ObjectToDelete) {
+	versioned := globalBucketVersioningSys.Enabled(bucket)
+	versionSuspended := globalBucketVersioningSys.Suspended(bucket)
+	for remaining := toDel; len(remaining) > 0; toDel = remaining {
+		if len(toDel) > maxDeleteList {
+			remaining = toDel[maxDeleteList:]
+			toDel = toDel[:maxDeleteList]
+		} else {
+			remaining = nil
+		}
+		deletedObjs, errs := o.DeleteObjects(ctx, bucket, toDel, ObjectOptions{
+			Versioned:        versioned,
+			VersionSuspended: versionSuspended,
+		})
+		for _, dobj := range deletedObjs {
```

Comment:
We only notify upon success @krisis - this is notifying blindly for all objects.

Reply:
done